ComponentKit/Core/Scope/CKThreadLocalComponentScope.mm

```diff
@@ -27,36 +27,46 @@ static pthread_key_t _threadKey() noexcept
   return thread_key;
 }
 
+static std::stack<CKThreadLocalComponentScope *> *_threadLocalStack() noexcept
+{
+  return static_cast<std::stack<CKThreadLocalComponentScope *> *>(pthread_getspecific(_threadKey()));
+}
+
 CKThreadLocalComponentScope *CKThreadLocalComponentScope::currentScope() noexcept
 {
-  return (CKThreadLocalComponentScope *)pthread_getspecific(_threadKey());
+  const auto threadLocalStack = _threadLocalStack();
+  return (threadLocalStack != nullptr && !threadLocalStack->empty()) ? threadLocalStack->top() : nullptr;
 }
 
 CKThreadLocalComponentScope::CKThreadLocalComponentScope(CKComponentScopeRoot *previousScopeRoot,
                                                          const CKComponentStateUpdateMap &updates)
 : newScopeRoot([previousScopeRoot newRoot]), stateUpdates(updates), stack()
 {
-  CKCAssert(CKThreadLocalComponentScope::currentScope() == nullptr, @"CKThreadLocalComponentScope already exists");
   stack.push({[newScopeRoot rootFrame], [previousScopeRoot rootFrame]});
   keys.push({});
-  pthread_setspecific(_threadKey(), this);
+
+  auto threadLocalStack = _threadLocalStack();
+  if (threadLocalStack == nullptr) {
+    threadLocalStack = new std::stack<CKThreadLocalComponentScope *>();
+    pthread_setspecific(_threadKey(), threadLocalStack);
```

Comment:
There is thread-local storage in C++, not sure if we can use it here. I've never used it: http://en.cppreference.com/w/cpp/keyword/thread_local